### ISSUE TITLE
The relationship between 'dim' and 'inner_dim'

### ISSUE THREAD

**Issue by HAN-oQo** [lucidrains/deformable-attention#2]
Hi, I have a question about  DeformableAttention module,

I calculated the output volumes of  the forward processes step by step,
According to my calculation, the code works only when 'dim' and 'inner_dim' is same. 

Is there any reason why you implement it this way?

Best regards,
Hankyu

**Comment by lucidrains** [comment]
@hanq0212  Hi Hankyu! Can you give an example of what you mean? Perhaps a code snippet?

**Comment by HAN-oQo** [comment]
Thank you for quick answer! 
for example,

```
x =  (32, 128, 224, 224)
DeformableAttention2D(dim=128, dim_head = 64, heads = 4, offset_sclae = 2, offset_group = 4)
# dim = 128, inner_dim = 64 * 4 = 256
```


forward_process:
```
# 32, 128, 224, 224 -> 32, 256, 224, 224
q = self.to_q(x)

#  32, 256, 224, 224 -> 128, 64, 224, 224
grouped_feats = rearrange(q, 'b (g d) ... -> (b g) d ...', g = self.offset_groups)

# 128, 64, 224, 224 -> 128, 2, 56, 56
offsets = self.to_offsets(grouped_feats)

# grid: 2, 56, 56 
grid =create_grid_like(offsets)

# vgrid: 128, 2, 56, 56 
vgrid = grid + offsets

# vgrid_scaled : 128, 56, 56, 2
vgrid_scaled = normalize_grid(vgrid)

# output kv_feats: 128, 64, 56, 56
kv_feats = F.grid_sample(
            grouped_feats,
            vgrid_scaled,
        mode = 'bilinear', padding_mode = 'zeros', align_corners = False)

# 128, 64, 56, 56  -> 32, 256, 56, 56
kv_feats = rearrange(kv_feats, '(b g) d ... -> b (g d) ...', b = b)

#  here ,
# .to_kv = nn.Conv2d(dim = 128, inner_dim = 256,  ...)
# input kv_feats : 32, 256, 56, 56
# input dimension is expected to 128

k, v = self.to_kv(kv_feats).chunk(2, dim = 1)
```

**Comment by lucidrains** [comment]
@hanq0212 oh yes, it does appear that the code doesn't account for the grouping correctly

do you want to try updating this to 0.0.10 and see if that works? your example above should run (albeit on a smaller feature map dimension of 32 x 32)

```python
import torch
from deformable_attention import DeformableAttention

attn = DeformableAttention(
    dim = 128,                   # feature dimensions
    dim_head = 64,               # dimension per head
    heads = 4,                   # attention heads
    dropout = 0.,                # dropout
    downsample_factor = 4,       # downsample factor (r in paper)
    offset_scale = 2,            # scale of offset, maximum offset
    offset_groups = 4,           # number of offset groups, should be multiple of heads
    offset_kernel_size = 6,      # offset kernel size
)
x = torch.randn(32, 128, 32, 32)
out = attn(x) 
assert (x.shape == out.shape)
```

**Comment by HAN-oQo** [comment]
Now it works well! Thank you
